python/taichi/lang/transformer.py

```diff
@@ -877,9 +877,49 @@ def visit_Return(self, node):
 class ASTTransformerChecks(ASTTransformerBase):
     def __init__(self, func):
         super().__init__(func)
+        self.has_return = False
+        self.in_static_if = False
 
     def visit_Call(self, node):
         if isinstance(node.func, ast.Name):
             node.args = [node.func] + node.args
             node.func = self.parse_expr('ti.func_call_with_check')
         return node
+
+    @staticmethod
+    def get_taichi_decorator(node):
+        if isinstance(node, ast.Call) and isinstance(
+                node.func, ast.Attribute) and isinstance(
+                    node.func.value, ast.Name) and node.func.value.id == 'ti':
+            attr = node.func.attr
+            if attr in ['static', 'grouped']:
+                return attr
+
+        return None
+
+    def visit_If(self, node):
```

Comment:
Looks like we already had a `visit_If` in this file?

Reply:
We're moving towards multipass AST transforming system. So this is actually the second pass.

Reply:
> Edit: OIC, this is the second pass. Do they need to be in the second pass?

Not necessary, but putting things too tight together increases maintainance cost. The `ASTTransformerChecks` is a pass focus on sanity check. Composing sanity logics to already-complicated-preprocessor logics sounds too painful, isn't it?